### ISSUE TITLE
[Feature] Requirements packages version lock

### ISSUE THREAD

**Issue by MatProv** [crim-ca/weaver#186]
**Is your feature request related to a missing implementation or functionality? Please describe.**
Missing package version lock.

**Describe the solution you'd like**
Idempotency between weaver installations.

**Describe alternatives you've considered**
Somehow lock python dependencies versions, to avoid undesired behaviours.

**Additional context**
Dependency management is crucial for this type of project. 
Having considerable amount of libraries that are always being upgraded to their latest version can guarantee that weaver's behaviours will change for unknown reasons. Simply running test suite isn't enough.

To put things in perspective, locking package versions will require extra effort to manually update these. 
However, it will take less time than investigating for a long period of time, only to realize that a library that was working two months ago has been updated last week and now breaks process execution.

**Comment by fmigneault** [comment]
Using the docker image should guarantee you to have the same packages it was built with for a given version. 

I think that if some *problematic* issue happens due to behavior change of a sub-package, and that it does not get caught, it is an indicator that some unittest is missing. Having flexible packages gives us quick bugfixes to many dependencies that are evolving in parallel to weaver, as many standards are employed here. 

I don't think we have sufficient resources to keep up with all changes.

**Comment by MatProv** [comment]
Yes I agree with you that this would involve time to manually update, and yes we want the latest bug fixes and standards.

Maybe not all the components need to have version lock but most of the crucial ones (eg: cwltool, celery, etc.).

At a higher level, take for example a docker-compose file. Would you prefer `weaver:latest` rather than `weaver:1.12.0`?
No, simply because `latest` would sometimes have undesired behaviours.
Of course this comparison target a single component (weaver, in the example) and not tens or hundreds, but the idea stays the same : idempotency.

That being said, locking version can't replace good tests nor pretend to avoid dependency issues.
I suggest this based on recent failures being reported as issues occurring from wrong package version, like [cwltool](https://github.com/crim-ca/weaver/blob/7353f4533f266c6e97a10b99ad4a4ee78a7bceba/requirements.txt#L17) and [celery](https://github.com/crim-ca/weaver/blob/b81353a35c66eb3e804a61984ae34de5c257c9fa/requirements.txt#L6).

**Comment by fmigneault** [comment]
> Maybe not all the components need to have version lock but most of the crucial ones (eg: cwltool, celery, etc.).

All packages that have some breaking changes are pinned accordingly.

As for the docker-compose, yes, I would actually prefer to have a staging environment that has `latest`. This way we always test the latest fixes and we can identify problems quicker. Of course, a prod deployment would have a specific image tag, but this is the same as a release version of this single component, a snapshot at that given time. 

In the case of the repo, `master` is simply our the equivalent of such latest-staging environment. It does not mean latest is guaranteed to not have errors or that new dependencies changes won't break. For this, one should use the last tagged version.
